internal/api/mfa.go

```diff
@@ -245,7 +284,34 @@ func (a *API) VerifyFactor(w http.ResponseWriter, r *http.Request) error {
 		return badRequestError("%v has expired, verify against another challenge or create a new challenge.", challenge.ID)
 	}
 
-	if valid := totp.Validate(params.Code, factor.Secret); !valid {
+	valid := totp.Validate(params.Code, factor.Secret)
+
+	if config.Hook.MFAVerificationAttempt.Enabled {
+		input := hooks.MFAVerificationAttemptInput{
+			UserID:   user.ID,
+			FactorID: factor.ID,
+			Valid:    valid,
+		}
+		output := &hooks.MFAVerificationAttemptOutput{}
+		response, err := a.invokeHook(ctx, input, output)
```

Comment:
Why do we need to return a response and error here? I'd prefer if it's just an error, as we pass `output` which will be filled on success (and even when there's an error).

This means you don't have to do casting below again, simplifying the code and removing the need to handle a ton of errors that just won't happen.

Reply:
I see where you're coming from and I'd also prefer to return just an error. The error was added with the intention of handling any issues which happen during unmarshalling which would produce an `error` but not an `output`.

Let me know if there's something I'm missing though